Next phase of command handler testing support (#414)

johnsonr opened the pull request:
Next phase of command handler testing support. Also:

- Improves duplicate removal in expression engine
- Makes TypeScript test stubs addressable (to help with deduping)
- Corrects some usages of `TreeNode` where `GraphNode` could be used
- Adds the concept of a `contextRoot` node to `RugContext` to enable handlers to run path expressions as needed. This change impacts runtime as well as testing, but is necessary to support our programming model.

Please squash on merge.

slimslenderslacks commented:
Each time we pass a message to rug-runner, we will _always_ include the team-id;  therefore, it's always possible to instantiate the ChatTeam root node whenever a CommandHandler or EventHandler runs.  So the requirement that a RugContext has a mandatory contextRoot is consistent with how we pass messages to rug.

ddgenome commented on src/main/scala/com/atomist/rug/runtime/js/JavaScriptCommandHandler.scala:
```diff
@@ -34,9 +34,6 @@ class JavaScriptCommandHandlerFinder
     * Extract intent from a var
-    *
-    * @param someVar
-    * @return
```
Any reason not to provide docs here? The scaladocs are publicly available now.

johnsonr replied:
I think we fill them in or delete them.

ddgenome commented on src/main/scala/com/atomist/rug/runtime/js/RugContext.scala:
```diff
@@ -44,4 +58,6 @@ object LocalRugContext extends RugContext {
     */
   override def teamId: String = "LOCAL_CONTEXT"
+
+  override def contextRoot(): GraphNode = ???
```
Does the not implemented here indicate it is not implemented yet or there is no valid implementation. If the latter, perhaps contextRoot does not belong in the parent class?

kipz replied:
Could we move LocalContext to test directory? I'm not sure it's used in cli/runner...

johnsonr replied:
I'd prefer to move it. What is `LocalContext` for?

ddgenome commented on src/main/scala/com/atomist/rug/runtime/js/interop/NashornMapBackedGraphNode.scala:
```diff
@@ -65,12 +66,23 @@ private class NashornMapBackedGraphNode(som: ScriptObjectMirror,
   }
 
+  override def address: String = relevantPropertiesAndValues.get(NodeAddressField) match {
+    case None => null
```
Can we not use an Option here or return an empty string as for nodeName?

johnsonr replied:
An empty string isn't an out of band value. It means the root of the path. I agree that using a possible `null` return is ugly. Think I may be able to fix this via a mixin in `NashornGraphNode`.

ddgenome commented on src/main/scala/com/atomist/rug/test/gherkin/handler/AbstractHandlerScenarioWorld.scala:
```diff
@@ -34,6 +39,17 @@ abstract class AbstractHandlerScenarioWorld(definitions: Definitions, rugs: Opti
+    val gn = NashornMapBackedGraphNode.toGraphNode(n).getOrElse(
+      throw new IllegalArgumentException(s"$n is not a valid GraphNode")
+    )
+    println(s"Adding to root context: $gn")
```
Do you want to leave the `println` in?

johnsonr replied:
Removed. Thanks

ddgenome commented on src/main/scala/com/atomist/tree/pathexpression/ExecutionResult.scala:
```diff
@@ -0,0 +1,50 @@
+    var nodeAddressesSeen: Set[String] = Set()
+    var toReturn = List.empty[GraphNode]
+    results.distinct.foreach {
+      case gn: AddressableGraphNode if gn.address != null =>
```
I see here you are testing against `null` for the address.

ddgenome replied:
It still would be good to avoid using `null` with an Option.

johnsonr replied:
We can now rely on the value being present, but I left the null check as a safety measure given that these values can come back from Nashorn. I know it's a bit ugly...

ddgenome commented on src/test/resources/com/atomist/rug/test/gherkin/handler/command/SingleMessageFeature1StepsWithMatchingPathExpression.ts:
```diff
@@ -0,0 +1,16 @@
+When("a visionary leader enters", (rugContext, world) => {
+   let handler = world.commandHandler("RunsMatchingPathExpressionCommandHandler")
+    let c = new node.Commit().withMadeBy(new node.Person("Ebony"))
+    world.addToRootContext(c)
```
indentation

ddgenome commented on src/test/resources/com/atomist/rug/test/gherkin/handler/event/PassingFeature1Steps.ts:
```diff
@@ -7,6 +7,6 @@ When("a visionary leader enters", world => {
    world.sendEvent(new node.Commit)
 })
-Then("excitement ensues", (p,world) => {
+Then("excitement ensues", world => {
```
I did not know this form of Then existed.

Lawouach replied:
Yeah, that's handy when you want to be explicit about the type:

```
(p: Project, world: SomeWorld)
```

johnsonr replied:
The second parameter is optional. It is always passed from the JVM

ddgenome replied:
What caught my eye was that when only one parameter is provided, it could be `world`. I've only ever seen the single parameter form have the one parameter be a `p: Project`.

ddgenome replied:
To be more clear: now what I am asking is whether line 10 should be

```typescript
Then("excitement ensues", world => {
```

or

```typescript
Then("excitement ensues", p => {
```

ddgenome replied:
It looks like `plan` is implemented on `HandlerScenarioWorld`, so perhaps it should remain `(p, world)`.

johnsonr replied:
`HandlerScenarioWorld` returns `world` for both parameters. The first parameter is really "fixture". In testing projects, you typically only care about the project, which is the fixture. In testing handlers, you care about the world, so it's both the fixture and itself.

kipz commented on src/main/scala/com/atomist/rug/runtime/js/interop/NashornMapBackedGraphNode.scala:
```diff
@@ -65,12 +66,23 @@ private class NashornMapBackedGraphNode(som: ScriptObjectMirror,
+      .toMap
+
   override def nodeName: String = relevantPropertiesAndValues.get(NodeNameField) match {
     case None => ""
```
Could we use Option here instead of ""?

johnsonr replied:
The value can now be relied on

kipz commented on src/main/scala/com/atomist/rug/runtime/js/interop/jsSafeCommittingProxy.scala:
```diff
@@ -175,7 +175,10 @@ class jsSafeCommittingProxy(
       nodesAccessedThroughThisFunctionCall.toList match {
         case Nil => throw new RugRuntimeException(name, s"No children or function found for property $name on $node")
         case null :: Nil => null
```
What does this mean?

johnsonr replied:
First case is an empty list. Second is a list with one null element. I don't remember where this is used, but it's not new.

kipz commented on src/main/scala/com/atomist/rug/test/gherkin/handler/AbstractHandlerScenarioWorld.scala:
```diff
@@ -17,7 +20,9 @@ abstract class AbstractHandlerScenarioWorld(definitions: Definitions, rugs: Opti
   private var planOption: Option[Plan] = None
```
Shouldn't `planOption` be `plan`. The type tells us this is an Option.

johnsonr replied:
`plan` is used to expose to JS, which won't like `Option`

kipz commented on src/main/scala/com/atomist/tree/pathexpression/ExecutionResult.scala:
```diff
@@ -0,0 +1,50 @@
+/**
+  * Result type and utility methods of a node navigation.
+  */
+object ExecutionResult {
```
Do we execute path expressions or evaluate them? I'm wondering if Execution is a little misleading.

johnsonr replied:
I agree that `EvaluationResult` might be better, but this will break the runner as it's a public API. Let's consider after this PR if we want to do that now.

kipz commented on src/main/scala/com/atomist/tree/pathexpression/ExecutionResult.scala:
```diff
@@ -0,0 +1,50 @@
+    * and under "/src/main" and under the actual file.
+    * So check that our returned types have distinct addresses
+    */
+  private def dedupe(results: Seq[GraphNode]): Seq[GraphNode] = {
```
I wonder if there's a more functional and less mutating way to do this.

In the past, I've used variants of this:

`list.groupBy(_.address).map(_._2.head)`

which seems appropriate here, perhaps with a filter too.

johnsonr replied:
It would be nicer for it to be more functional, but it's not trivial, given the fact that not all nodes may have an address and we need to preserve overall order.

kipz commented on src/main/typescript/node_modules/@atomist/rug/tree/PathExpression.ts:
```diff
@@ -72,6 +72,14 @@ interface TreeNode extends GraphNode {
+    /**
+     * Path from root of graph or tree
+     */
+    address(): String
```
should be lower case `string`

johnsonr replied:
Thanks. Fixed

kipz commented on src/test/resources/com/atomist/rug/test/gherkin/handler/command/SingleMessageFeature1Steps.ts:
```diff
@@ -0,0 +1,13 @@
+})
+Then("excitement ensues", world => {
+   console.log("The plan message were " + world.plan().messages())
+    return world.plan().messages().length == 1
```
indentation.

ddgenome replied:
I have a cleanup of the indentation and semicolons in a branch.

https://github.com/atomist/rug/commit/ddba862ba474934ae1aa7d9264ff9705379506ae

kipz commented on src/main/typescript/node_modules/@atomist/rug/tree/TreeHelper.ts:
```diff
@@ -66,3 +66,15 @@ export class TreeHelper {
+
+export abstract class AddressedNodeSupport implements Addressed {
+
+    private _address: string = null
```
I think `private _address: string;` is fine.

johnsonr replied:
Changed

johnsonr replied:
This code ended up changed quite a bit anyway, as I found and fixed a bug in address handling

kipz commented on src/main/typescript/node_modules/@atomist/rug/tree/TreeHelper.ts:
```diff
@@ -66,3 +66,15 @@ export class TreeHelper {
     }
 }
+
+export abstract class AddressedNodeSupport implements Addressed {
```
AFAIK, it's idiomatic to add semicolons at the end of line.

johnsonr replied:
Maybe it's all the Scala, but semicolons just add noise IMO. Also, the vast majority of our generated and library TS code doesn't use them. Either we change everywhere or not.

kipz commented:
I wouldn't be comfortable squashing somebody else commits TBH. IMHO - the author should do this.

johnsonr commented:
I think I've addressed the comments, besides those which I think need further discussion and should be handled separately.

Also fixed a bug in address handling.
